services/sidecar/src/simcore_service_sidecar/executor.py

```diff
@@ -0,0 +1,353 @@
+import json
+import logging
+import shutil
+import time
+from pathlib import Path
+from typing import Dict
+
+import aiodocker
+import aiopg
+import attr
+from celery.utils.log import get_task_logger
+from packaging import version
+
+from servicelib.utils import fire_and_forget_task, logged_gather
+from simcore_sdk import node_data, node_ports
+from simcore_sdk.node_ports.dbmanager import DBManager
+
+from . import config, exceptions
+from .log_parser import LogType, monitor_logs_task
+from .rabbitmq import RabbitMQ
+
+log = get_task_logger(__name__)
+
+
+@attr.s(auto_attribs=True)
+class TaskSharedVolumes:
+    input_folder: Path = None
+    output_folder: Path = None
+    log_folder: Path = None
+
+    @classmethod
+    def from_task(cls, task: aiopg.sa.result.RowProxy):
+        return cls(
+            Path.home() / f"input/{task.job_id}",
+            Path.home() / f"output/{task.job_id}",
+            Path.home() / f"log/{task.job_id}",
+        )
+
+    def create(self) -> None:
+        for folder in [
+            self.input_folder,
+            self.output_folder,
+            self.log_folder,
+        ]:
+            if folder.exists():
+                shutil.rmtree(folder)
+            folder.mkdir(parents=True, exist_ok=True)
+
+
+@attr.s(auto_attribs=True)
+class Executor:
+    db_engine: aiopg.sa.Engine = None
+    db_manager: DBManager = None
+    rabbit_mq: RabbitMQ = None
+    task: aiopg.sa.result.RowProxy = None
+    user_id: str = None
+    stack_name: str = config.SWARM_STACK_NAME
+    shared_folders: TaskSharedVolumes = None
+    integration_version: version.Version = version.parse("0.0.0")
+
+    async def run(self):
+        log.debug(
+            "Running %s project:%s node:%s internal_id:%s from container",
+            self.task.image["name"],
+            self.task.project_id,
+            self.task.node_id,
+            self.task.internal_id,
+        )
+        try:
+            await self.preprocess()
+            await self.process()
+            await self.postprocess()
+            await self._post_messages(
+                LogType.LOG, "[sidecar]...task completed successfully."
+            )
+        except exceptions.SidecarException:
+            await self._post_messages(LogType.LOG, "[sidecar]...task failed.")
+            raise
+
+    async def preprocess(self):
+        await self._post_messages(LogType.LOG, "[sidecar]Preprocessing...")
+        log.debug("Pre-Processing...")
+        self.shared_folders = TaskSharedVolumes.from_task(self.task)
+        self.shared_folders.create()
+        await logged_gather(self._process_task_inputs(), self._pull_image())
+        log.debug("Pre-Processing Pipeline DONE")
+
+    async def process(self):
+        log.debug("Processing...")
```

Comment:
IDEA: we could implement a small context manager  to log these steps
```python
import logging
from servicelib.log_utils import StepLogger

log = logging.getLogger(__name__)
log_step = StepLogger(log)

async def process(self):
     with log_step.debug("Doing X"):
         foo()
         log.debug("yes")
         bar()
     with log_step.debug("Doing Y"):
         zee()
         bla()
```
would produce
```
DEBUG: Doing X ... [started]
DEBUG: yes
DEBUG: Doing X [done]
DEBUG: Doing Y[started]
DEBUG: Doing Y [dome]
```

can even be extended as decorator as well so we can log start and end of functions.